### ISSUE TITLE
How to stop processing new requests while we're still streaming results from earlier ones?

### ISSUE THREAD

**Issue by kivikakk** [facebook/proxygen#240]
Hi! I'm working with Proxygen to make a simple REST-ish server. Here's the context:

* simple server that runs with number of threads = number of processors * cores.
* `onEOM` we perform a CPU-intensive operation on the input data and construct a response. (for a ~1mb input we produce ~8mb of output; the computation itself takes about ~2s of CPU time.)

The operation produces output as a stream; at first, I was wiring this through to the `ResponseBuilder` directly, sending the response through in chunks to limit memory pressure. To my surprise, this was many times (!) slower than just accumulating the response document in an `std::string` and then handing it to `ResponseBuilder::body` in one hit. Peak memory use is probably higher but I'm not too concerned by that itself.

The issue is with a large number of concurrent requests. Here's what appears to happen:

* initiate 100 identical requests against the server.
* over the next 4 seconds, 100 of my `RequestHandler`s are constructed.
* sure enough, only 8 handlers are actually receiving callbacks at any one time.
* when a handler is finished handing the response back off to Proxygen (i.e. we've called `builder.body(blah).sendWithEOM();` and have nothing else to do), other handlers get to work on the now-idle threads.
* at this point in time, the first requests — while "handled" from my point of view — are still sending the 8mb of response data to the clients.
* further handlers complete, meaning more responses are simultaneously being sent. (again, this is all out of the hands of my handlers' callbacks — this appears to be happening in a different thread. this leads me to assume there's a thread or two dedicated to evented IO which is actually sending response data to clients.)
* a few of the early responses will actually make it through, but the responses-in-flight accumulate — by the end of the process, all 100 computations are finished and handed back off to Proxygen, but there's still 46 active streams.
* over the following seconds, those streams finally finish draining.

In short, we don't prioritise finishing sending data of earlier requests, meaning as concurrent requests increase, *all* responses slow down, despite a fixed workload at any point in time. If requests happened to be staggered, we'd serve everyone in a timely fashion, but if a large number come in all at once, then despite processing them roughly in order, *all* requests take a long time to complete.

I'm wondering if there's anything I can do about this — ideally we wouldn't start processing a 9th request until the 1st has finished streaming its response. Or we'd process it, but wouldn't try streaming its response until we're down to 7 simultaneous responses being streamed. Or _something_.

I'd be super grateful for any ideas or comments! :heart:

**Comment by afrind** [comment]
That's the best way to do it.

**Comment by jadhavabhi** [comment]
Hi, 
I am encountering a similar issue in my application where I am using Proxygen HTTP threads as worker threads and Folly IO threads as consumer threads. The configuration includes the default HTTP server options, with a thread count of 240 for Proxygen and 384 for the IO worker threads.

Under high load, I observe that some HTTP threads are handling more than one request simultaneously, reaching up to 8 concurrent requests, while other HTTP threads are not fully utilized. This uneven load distribution is resulting in increased response times, as requests are placed in a wait state.

Additionally, I’ve noticed a significant backlog of gzip-compressed requests, and the load balancer seems to be directing more load to these specific threads.

Could you please advise on how to configure the system to limit the number of concurrent requests handled by each HTTP thread to a maximum of 3? Moreover, I’d like to prevent the accumulation of a specific type of request in the backlog.

Any insights or suggestions would be greatly appreciated.

Thank you.
